test/cluster/basic_test.go

```diff
@@ -101,3 +103,424 @@ func TestBasicCluster(t *testing.T) {
+
+	var clusterStatusResEnv *types.GetClusterStatusResponseEnvelope
+	leaders := [3]string{"", "", ""}
+	for i, srv := range c.Servers {
+		require.Eventually(t, func() bool {
+			clusterStatusResEnv, err = srv.QueryClusterStatus(t)
+			return clusterStatusResEnv != nil && clusterStatusResEnv.GetResponse().GetLeader() != "" && err == nil
+		}, 30*time.Second, 100*time.Millisecond)
+		leaders[i] = clusterStatusResEnv.GetResponse().GetLeader()
+	}
+
+	require.Equal(t, leaders[0], leaders[1])
+	require.Equal(t, leaders[0], leaders[2])
+
+	leaderServer, leaderVal := c.GetServerByID(leaders[0])
+	require.NotEqual(t, leaderVal, -1)
+	require.NotNil(t, leaderServer)
+
+	require.NoError(t, c.ShutdownServer(leaderServer))
+
+	//find the new leader
+	var newLeader int
+	for _, srv := range c.Servers {
+		if srv != leaderServer {
+			require.Eventually(t, func() bool {
+				clusterStatusResEnv, err = srv.QueryClusterStatus(t)
+				return clusterStatusResEnv != nil && clusterStatusResEnv.GetResponse().GetLeader() != "" && err == nil
+			}, 30*time.Second, 100*time.Millisecond)
+			_, newLeader = c.GetServerByID(clusterStatusResEnv.GetResponse().GetLeader())
+		}
+	}
+
+	txID, rcpt, err := c.Servers[newLeader].WriteDataTx(t, worldstate.DefaultDBName, "bob", []byte{uint8(newLeader)})
+	require.NoError(t, err)
+	if err != nil {
+		t.Logf("txID: %s, error: %s", txID, err)
+	} else {
+		t.Logf("tx submitted: %s, %+v", txID, rcpt)
+	}
+
+	require.NoError(t, c.StartServer(leaderServer))
+
+	var dataEnv *types.GetDataResponseEnvelope
+	require.Eventually(t, func() bool {
+		dataEnv, err = leaderServer.QueryData(t, worldstate.DefaultDBName, "bob")
+		return dataEnv != nil && dataEnv.GetResponse().GetValue() != nil && err == nil
+	}, 30*time.Second, 100*time.Millisecond)
+
+	dataResp := dataEnv.GetResponse().GetValue()
+	require.Equal(t, dataResp, []byte{uint8(newLeader)})
+	t.Logf("data: %+v", dataResp)
+}
+
+// Scenario:
+// - start 3 servers in a cluster.
+// - wait for one to be the leader.
+// - submit 2 txs
+// - shutting down and starting all servers in the cluster.
+// - make sure the servers is in sync with the txs made before they were shut down.
+func TestClusterRecovery(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	//find the leader
+	var clusterStatusResEnv *types.GetClusterStatusResponseEnvelope
```

Comment:
The check that all servers agree on a leader should be part of the "Eventually", that is you poll untill they all give a consistent answer. See for example [this function ](https://github.com/hyperledger-labs/orion-server/blob/67a69d1a6e0bb46a53590e1538a37d7417f8d35b/internal/replication/env_test.go#L345) and how  it is used in the replication unit tests.